handshake/src/types/covenant.rs

```diff
@@ -0,0 +1,267 @@
+    pub covenant_type: CovenantType,
+    /// The CovenantData represents Covenant specific data that is attached to
+    /// the UTXO. This data must be updated correctly as the UTXO is spent from
+    /// one CovenantType to the next CovenantType.
+    pub covenant_data: CovenantData
+}
+
+impl Covenant {
+    /// Returns the null Covenant.
+    pub fn null() -> Self {
+        Self {
+            covenant_type: CovenantType::NONE,
+            covenant_data: CovenantData::null()
+        }
+    }
+}
+
+impl ByteFormat for Covenant {
+    type Error = SerError;
+
+    fn serialized_length(&self) -> usize {
+        let mut size: usize = 1;  // covenant_type
+        size += self.covenant_data.serialized_length();
+        size
+    }
+
+    fn read_from<R>(reader: &mut R, _limit: usize) -> SerResult<Self>
+    where
+        R: Read,
+        Self: std::marker::Sized,
+    {
+
+        let mut buf = [0u8; 1];
+        reader.read_exact(&mut buf)?;
+        let covenant_type = u8::from_le_bytes(buf);
+        // TODO(mark): sane max?
+        let covenant_data = CovenantData::read_from(reader, 1024)?;
+
+        Ok(Self {
+            covenant_type: covenant_type.into(),
+            covenant_data
+        })
+    }
+
+    fn write_to<W>(&self, writer: &mut W) -> SerResult<usize>
+    where
+        W: Write,
+    {
+        let mut total: usize = 0;
+        let covenant_type: u8 = self.covenant_type.clone() as u8;
+        total += writer.write(&[covenant_type])?;
+        total += &self.covenant_data.write_to(writer)?;
+
+        Ok(total)
+    }
+}
+
+
+/// Covenant types
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq)]
+#[repr(u8)]
+pub enum CovenantType {
+    /// Standard Bitcoin-like Output
+    NONE = 0,
+    /// Reserved Name Claim Output
+    CLAIM = 1,
+    /// Auction Opening Output
+    OPEN = 2,
+    /// Blinded Bid Output
+    BID = 3,
+    /// Reveal Output
+    REVEAL = 4,
+    /// Auction Losers Must Redeem
+    REDEEM = 5,
+    /// Auction Winners Must Register
+    REGISTER = 6,
+    /// Update Name State
+    UPDATE = 7,
+    /// Renew Name
+    RENEW = 8,
+    /// Begin to Transfer Name
+    TRANSFER = 9,
+    /// Finalize Name Transfer
+    FINALIZE = 10,
+    /// Burn Name When Key is Compromised
+    REVOKE = 11,
+    /// Unknown Type
+    UNKNOWN = 255
+}
+
+impl From<u8> for CovenantType {
+    fn from(u: u8) -> Self {
+        match u {
+            0x00 => CovenantType::NONE,
+            0x01 => CovenantType::CLAIM,
+            0x02 => CovenantType::OPEN,
+            0x03 => CovenantType::BID,
+            0x04 => CovenantType::REVEAL,
+            0x05 => CovenantType::REDEEM,
+            0x06 => CovenantType::REGISTER,
+            0x07 => CovenantType::UPDATE,
+            0x08 => CovenantType::RENEW,
+            0x09 => CovenantType::TRANSFER,
+            0x0a => CovenantType::FINALIZE,
+            0x0b => CovenantType::REVOKE,
+            _ => CovenantType::UNKNOWN
```

Comment:
This needs to be updated so that it catches unknown covenant types and passes the byte through instead of using `UNKNOWN` as `255`. This could potentially break things. Covenants are supposed to be able to be soft forked in.